tests/test_server.py

```diff
@@ -545,3 +545,41 @@ def cmd(self):
         with pytest.raises(DevFailed) as record:
             proxy.cmd()
         assert "ZeroDivisionError" in record.value.args[0].desc
+
+# Test server dynamic attributes async mode
+
+def test_dyn_attr_async_read(typed_values, server_green_mode):
+    py_dtype, values, expected = typed_values
+
+    from tango.device_proxy import __get_tango_type
+    dtype = __get_tango_type(py_dtype)
+
+    class TestDevice(Device):
+        green_mode = server_green_mode
+        _is_testattr_allowed = True
+
+        def __init__(self, *args, **kwargs):
+            super(TestDevice, self).__init__(*args, **kwargs)
+
+            attr = Attr("TestAttr", dtype, AttrWriteType.READ_WRITE)
+            self.add_attribute(attr, self.read_attr, self.write_attr, self.is_attr_allowed)
+
+        def read_attr(self, attr):
+            attr.set_value(self.attr_value)
+
+        def write_attr(self, attr):
+            self.attr_value = attr.get_write_value()
+
+        def is_attr_allowed(self, req_type):
+            return self._is_testattr_allowed
+
+    with DeviceTestContext(TestDevice) as proxy:
+        proxy._is_testattr_allowed = True
```

Comment:
The `proxy` object is an instance of `DeviceProxy` connected to the device being tested.  The `DeviceTestContext` doesn't give us direct access to the Tango device it runs.  This means that line 577 isn't modifying `TestDevice._is_testattr_allowed` (or the instance variable for the device under test), but rather adding a new attribute to the the `proxy` object. 

Add a static command or read/write attribute to `TestDevice` and use that to toggle the internal variable.